crates/ruff_linter/src/rules/pylint/rules/bad_string_format_character.rs

```diff
@@ -90,27 +89,10 @@ pub(crate) fn call(checker: &mut Checker, string: &str, range: TextRange) {
 
 /// PLE1300
 /// Ex) `"%z" % "1"`
-pub(crate) fn percent(checker: &mut Checker, expr: &Expr) {
-    // Grab each string segment (in case there's an implicit concatenation).
-    let mut strings: Vec<(TextRange, AnyStringFlags)> = vec![];
-    for (tok, range) in
-        lexer::lex_starts_at(checker.locator().slice(expr), Mode::Module, expr.start()).flatten()
-    {
-        match tok {
-            Tok::String { flags, .. } => strings.push((range, flags)),
-            // Break as soon as we find the modulo symbol.
-            Tok::Percent => break,
-            _ => {}
-        }
-    }
-
-    // If there are no string segments, abort.
-    if strings.is_empty() {
-        return;
-    }
-
-    for (range, flags) in &strings {
-        let string = checker.locator().slice(*range);
+pub(crate) fn percent(checker: &mut Checker, expr: &Expr, format_string: &ExprStringLiteral) {
+    for string_literal in &format_string.value {
+        let string = checker.locator().slice(string_literal);
+        let flags = AnyStringFlags::from(string_literal.flags);
```

Comment:
Just a thought that we could provide a trait which has common methods for all `StringLiteralFlags`, `BytesLiteralFlags`, and `FStringFlags` and other methods which needs to be implemented for concrete types so that we can use the same functionality for any kind of string flags.

Reply:
I think I had something like this in an early iteration of one of the string-flags PRs, but abandoned it. I think the reason I put it aside was because you need to have the trait imported for the trait methods to be available, so I felt like the methods would be hard for people implementing a new linter rule to discover. But I'm not opposed to something like that, if we think it could simplify things!

Reply:
I think you're right that it's unfortunate that we have to convert the `StringLiteralFlags` object to the more general `AnyStringFlags` type here in order to get the `closer_len()` method. That should really be available on `StringLiteralFlags`, `BytesLiteralFlags`, etc., as well as `AnyStringFlags`. So I agree that a trait would make sense here, especially now we're trying to reduce the number of linter rules that depend on the tokenizer

Reply:
Yeah, not a priority but something which I thought would be useful.